### DISCUSSION TITLE
Postgres : can't get upsert with entity properties using UnderscoreNamingStrategy to generate proper statement

### DISCUSSION THREAD

**Discussion by bsoulier** [mikro-orm/mikro-orm#4140] [Q&A]
**Describe the bug**
Upsert does not handle underscore casing of properties for Postgres driver.

**Stack trace**
```
error: insert into "scenarios" ("dueDate", "id", "name", "scenario", "type") values ('2023-03-15T20:40:20Z', '90877c24-c371-11ed-afa1-0242ac120002', 'dssdfsdfsdf', '49df3022-3547-49d3-a37e-ea16b5758cd5', 'balance') on conflict ("id") do update set "name" = excluded."name", "dueDate" = excluded."dueDate", "type" = excluded."type", "scenario" = excluded."scenario" returning "id" - column "dueDate" of relation "scenarios" does not exist
```

**To Reproduce**
1. Have an entity with fields such as dueDate
2. use a repository to perform an upsert or an upsertMany
3. error shows up

**Expected behavior**
property generated in sql statment to be 'due_date' instead of 'dueDate'.

**Additional context**

**Versions**

| Dependency | Version |

| node | v16.19.0 |
| typescript | 4.7.4 |
| mikro-orm | 5.6.13 |
| your-driver | postgres |

**Comment by B4nan** [comment]
Please don't expect everything to be so broken that it is enough to just point a finger - provide code so it's clear what you are doing.

Upserting is tested in all drivers the very same way (literally through the exact same code), so yes, it does work in postgres, especially with the default `UnderscoreNamingStrategy`.

https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/upsert/upsert.test.ts

I just tried to modify that test to use a property name `dueDate`, and it works as expected, producing queries with `due_date`.

edit: modified the test so you can see yourself, it now uses `propName` which translates to `prop_name` in the queries, which you can see in the snapshot:

https://github.com/mikro-orm/mikro-orm/blob/master/tests/features/upsert/__snapshots__/upsert.test.ts.snap

**Comment by bsoulier** [comment]
Hello @B4nan,

I am sorry about my issue being posted without enough details or posting from my part; was definitely an issue on my end, using the wrong @InjectRepository.

I will do my best to give more reproducible details next time, sorry for making you lose time on this.

**Reply by B4nan** [reply]
Can you provide more details about what was it? If you used the wrong repository, it sounds weird that it would generate a query for the correct/expected table instead of respecting the repository type.

**Reply by bsoulier** [reply]
The issue was with the setup occurring in my service constructor:
` @InjectRepository(Scenario) scenariosRowsRepository: ScenarioRowsRepository,`

The Inject annotation type and the destination repo were not matching, but the repository was still getting created.
Upserts were failing because the destination table was not matching columns, that were also not translated because not part of the right entity.

**Reply by B4nan** [reply]
I see, we had a recent discussion in #3989 that is related to this problem, it would be good to validate the repository methods that work with entity instances to ensure the type is same as the repository - same issue comes with `repo.assign` or `repo.populate`, if you make this mistake, things will type-check and we have no runtime validation there. Will try to implement that for v6.

**Reply by bsoulier** [reply]
That sounds wonderful! Thanks for letting me know, let me know if I could help
